### PR TITLE
DrivAerML: 2-head vs 16-head attention sweep on DM champion

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

At 512d hidden, the current DM champion uses 8 attention heads (64d/head). Noam's experiments on the AirfRANS architecture found that 3 heads were optimal at 192d (64d/head), suggesting per-head dimensionality matters. We want to understand whether the current 8H/64d-per-head choice is optimal for DrivAerML at 512d.

Valid head counts for 512d: 1, 2, 4, 8, 16, 32, 64, ... (must divide 512 evenly). In-flight coverage: senku #3165 tests 4 heads (128d/head), griffith #3160 tests 16 heads. This PR fills the remaining gap with **2 heads (256d/head)** — the widest-attention regime not yet tested. We also include **16 heads (32d/head)** as Trial 2 in case griffith #3160 was run without the EMA+gc stability combo required for DrivAerML convergence.

The hypothesis: wider per-head attention (fewer, larger heads) may be better suited to 3D aerodynamic surface point clouds where each attention head needs to capture long-range global pressure gradients, as opposed to the narrow multi-head regime more common in NLP.

## Instructions

Keep all other DM champion hyperparameters exactly as the baseline. Only vary `--model-heads`. Use `--wandb_group dm-heads-sweep` so all head-count runs (including senku and griffith) are visible together.

**Trial 1 — 2 heads (256d/head, very wide attention):**

```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 2 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --wandb_group dm-heads-sweep
```

**Trial 2 — 16 heads (32d/head, narrow attention):**

```bash
cd target/icml2026
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 16 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --wandb_group dm-heads-sweep
```

Run Trial 1 first; if it beats baseline, that's the result to report. Run Trial 2 after (or in parallel on a second GPU if available). Report both results with W&B run IDs.

**Important:** Both trials must use `--ema-decay 0.9995` together with `--grad-clip 0.5`. EMA alone (without gc=0.5) diverges on DrivAerML around epoch 28. Do not remove either flag.

## Baseline

Current DrivAerML best (PR #3072, eren/dm-ema-9995-gc05):

| Metric | Value |
|--------|-------|
| val_primary/surface_rel_l2_pct | **3.833%** (epoch 511) |
| test/surface_rel_l2_pct | 4.685% |
| Config | 4L / 512d / **8H** / AdamW lr=5e-4 / T_max=30 / EMA=0.9995 / gc=0.5 / Fourier / no WD |

- W&B run: `ncl1dh88` (eren/dm-ema-9995-gc05)
- External target: AB-UPT <3.71% (gap: 0.123 pp)
- Reproduce:
  ```bash
  cd target/icml2026
  SENPAI_MAX_EPOCHS=9999 python train.py \
    --dataset drivaerml --optimizer adamw --lr 5e-4 \
    --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
    --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
    --epochs 999 --batch-size 1 \
    --drivaerml-train-surface-points 50000 \
    --drivaerml-eval-surface-points 50000 \
    --max-train-batches 394 --max-eval-batches 200 \
    --ema-decay 0.9995
  ```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT chihiro:

## Results

### Trial 1 — 2 heads (256d/head, very wide attention)

| Metric | Value | Baseline (8H) | Delta |
|--------|-------|---------------|-------|
| val_primary/surface_rel_l2_pct | **13.878%** (ep35) | 3.833% (ep511) | +10.045 pp |
| test/surface_rel_l2_pct | 14.373% | 4.685% | +9.688 pp |

- W&B run: `qcaxbz95` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/qcaxbz95))
- Total epochs: 505 (hit timeout)
- **Catastrophic divergence at ep40.** Gradient norms exploded to Infinity; 394/394 grad clips every epoch thereafter. EMA preserved best model at ep35 but it was already very poor. The raw model immediately collapsed to ~75.86% val (random-level) and never recovered.
- Peak memory: not tracked in wandb

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 2 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --wandb-group dm-heads-sweep --agent chihiro \
  --wandb-name "chihiro/dm-2heads-256dphead"
```

### Trial 2 — 16 heads (32d/head, narrow attention)

| Metric | Value | Baseline (8H) | Delta |
|--------|-------|---------------|-------|
| val_primary/surface_rel_l2_pct | **4.099%** (ep428) | 3.833% (ep511) | +0.266 pp |
| test/surface_rel_l2_pct | 4.769% | 4.685% | +0.084 pp |

- W&B run: `6n9i88a3` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/6n9i88a3))
- Total epochs: 498 (hit timeout; NaN from ep433 onward)
- Remarkably stable for 430 epochs with near-zero grad clip events and steadily decreasing grad norms (0.25 → 0.09).
- Periodic perturbations (loss spikes every ~30-50 epochs) were absorbed by EMA with 0.9995 decay, though washout took ~15-20 epochs each time.
- **Eventually diverged at ep431** (grad_norm: 46.8) → ep432 (grad_norm: 4173.9) → ep433 (NaN). EMA preserved best checkpoint at ep428.
- Peak memory: not tracked in wandb

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 16 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --wandb-group dm-heads-sweep --agent chihiro \
  --wandb-name "chihiro/dm-16heads-32dphead"
```

### What happened

**Neither alternative beats the baseline 8-head configuration.** Both 2-head and 16-head eventually diverge, while the 8-head baseline ran stably to ep511.

The stability gradient across head counts reveals a clear pattern:

| Heads | d/head | Best Val | Diverged At | Stable? |
|-------|--------|----------|-------------|---------|
| 2 | 256 | 13.878% | ep40 | No — catastrophic |
| 8 | 64 | 3.833% | Never (ep511+) | Yes |
| 16 | 32 | 4.099% | ep431 | Mostly — late instability |

- **2-head (256d/head) is fundamentally unstable** for DrivAerML. With only 2 attention heads, each head must attend to an extremely high-dimensional space. A single bad gradient in one head can poison the entire attention mechanism, and with no redundancy from other heads, recovery is impossible. The divergence at ep40 was immediate and total.

- **16-head (32d/head) is competitive but fragile.** It trained well for 430 epochs, producing a strong 4.099% val — only 0.27pp behind baseline. The smaller per-head dimension produced more stable individual gradient updates (fewer clip events, lower grad norms). But the reduced representational capacity per head meant it couldn't match the 8-head's absolute performance, and the system eventually accumulated enough instability to diverge at ep431.

- **8 heads (64d/head) is the sweet spot** for this 4L/512d architecture on DrivAerML. It balances sufficient per-head representational capacity with enough head redundancy for gradient stability.

The hypothesis that wider per-head attention would be better for 3D aerodynamic point clouds is **refuted** — the opposite appears true. The 8-head baseline already sits at a good balance point.

### Suggested follow-ups

1. **Investigate the periodic perturbations.** Both configurations (and presumably the baseline) show loss spikes every ~30-50 epochs. These appear data-driven (possibly specific training samples). Understanding their source could improve stability across all configurations.
2. **Test 4 heads (128d/head)** — already in-flight with senku #3165, but worth comparing stability profile. Prediction: likely less stable than 8-head but more stable than 2-head.
3. **Consider lower learning rate for extreme head counts.** The 2-head divergence might be mitigated with lr=1e-4 or lr=2e-4, though even if stable the representational capacity concern remains.

### Bug fix note

Fixed a `primary_metric_key` bug in train.py where a local string variable shadowed the module-level function, causing an `UnboundLocalError` at the start of training. Consolidated duplicate initialization blocks (net -42/+18 lines). This fix was necessary for both trials to run.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3188 because ADVISOR: 2H=13.878% crashed ep40 (W&B: qcaxbz95). 16H=4.099% diverged ep431 (W&B: 6n9i88a3). Neither beats 3.833%. 8H (64d/head) is confirmed optimal. Now ALL head counts tested: 2H (catastrophic), 4H (6.650% diverged), 8H (3.833% champion), 16H (4.099% late-diverge). 8H provides the right balance — 2H/4H have too few heads for stable multi-head attention, 16H has too small per-head dim (32d) for expressive queries. Adding to dead ends..
